Source/WebCore/editing/cocoa/TextIteratorHTMLConverter.mm

```diff
@@ -0,0 +1,346 @@
+        CachedResourceHandle image = renderImage->cachedImage();
+        if (image && !image->errorOccurred()) {
+            RetainPtr<NSFileWrapper> wrapper = adoptNS([[NSFileWrapper alloc] initRegularFileWithContents:(__bridge NSData *)image->imageForRenderer(renderer)->adapter().tiffRepresentation()]);
+            [wrapper setPreferredFilename:@"image.tiff"];
+            return wrapper;
+        }
+    }
+
+    return nil;
+}
+
+static RetainPtr<NSFileWrapper> fileWrapperForElement(const HTMLAttachmentElement& element)
+{
+    auto identifier = element.uniqueIdentifier();
+
+    RetainPtr data = [(NSString *)identifier dataUsingEncoding:NSUTF8StringEncoding];
+    if (!data)
+        return nil;
+
+    // Use a filename prefixed with a sentinel value to indicate that the data is corresponding
+    // to an existing HTMLAttachmentElement.
+
+    RetainPtr wrapper = adoptNS([[NSFileWrapper alloc] initRegularFileWithContents:data.get()]);
+    [wrapper setPreferredFilename:makeString(WebContentReader::placeholderAttachmentFilenamePrefix, identifier)];
+    return wrapper;
+}
+
+static RetainPtr<NSAttributedString> attributedStringWithAttachmentForElement(const HTMLImageElement& element)
+{
+#if ENABLE(MULTI_REPRESENTATION_HEIC)
+    if (element.isMultiRepresentationHEIC()) {
+        if (RefPtr image = element.image()) {
+            if (WebMultiRepresentationHEICAttachment *attachment = image->adapter().multiRepresentationHEIC()) {
+                RetainPtr attachmentString = adoptNS([[NSString alloc] initWithFormat:@"%C", static_cast<unichar>(NSAttachmentCharacter)]);
+                RetainPtr attributedString = adoptNS([[NSMutableAttributedString alloc] initWithString:attachmentString.get()]);
+                [attributedString addAttribute:WebMultiRepresentationHEICAttachmentAttributeName value:attachment range:NSMakeRange(0, 1)];
+                return attributedString;
+            }
+        }
+    }
+#endif
+
+    RetainPtr fileWrapper = fileWrapperForElement(element);
+    RetainPtr attachment = adoptNS([[PlatformNSTextAttachment alloc] initWithFileWrapper:fileWrapper.get()]);
+    return [NSAttributedString attributedStringWithAttachment:attachment.get()];
+}
+
+static RetainPtr<NSAttributedString> attributedStringWithAttachmentForElement(const HTMLAttachmentElement& element)
+{
+    RetainPtr fileWrapper = fileWrapperForElement(element);
+    RetainPtr attachment = adoptNS([[PlatformNSTextAttachment alloc] initWithFileWrapper:fileWrapper.get()]);
+    return [NSAttributedString attributedStringWithAttachment:attachment.get()];
+}
+
+static bool hasAncestorQualifyingForWritingToolsPreservation(Element* ancestor, HashMap<Element*, bool>& cache)
+{
+    if (!ancestor)
+        return false;
+
+    return cache.ensure(ancestor, [&] {
+        if (isMailBlockquote(*ancestor))
+            return true;
+
+        // FIXME: This should use the render style of the element instead.
```

Comment:
need to figure out how to do this

Reply:
`renderer()->style()`?

Reply:
I specifically meant, how can I tell if it is a pre from the render style, not how to access the style.

Reply:
whiteSpace() == WhiteSpace::Pre